### PR TITLE
Allow for expedited nokeep array broadcasts

### PR DESCRIPTION
**Pull request by epmikida**

### CODE REVIEW

File: src/ck-core/ckarray.C [hunk 1]
```diff
@@ -1411,7 +1411,9 @@ void CProxy_ArrayBase::ckBroadcast(CkArrayMessage* msg, int ep, int opts) const
     if (CkMyPe() == CpvAccess(serializer)) {
       DEBB((AA "Sending array broadcast\n" AB));
       // TODO: What if we have a message thats nokeep AND expedited?
```
Comment (stwhite91) [inline src/ck-core/ckarray.C]: Can remove the TODO now

**Comment by evan-charmworks** [comment]
netlrts aborted in megampi (`./charmrun ./pgm +p2 +vp2  ++local +isomalloc_sync`) with `AMPI Logic error: received late out-of-order message!`

**Comment by stwhite91** [comment]
> netlrts aborted in megampi (`./charmrun ./pgm +p2 +vp2 ++local +isomalloc_sync`) with `AMPI Logic error: received late out-of-order message!`

I'll take a look

**Comment by epmikida** [comment]
> > netlrts aborted in megampi (`./charmrun ./pgm +p2 +vp2 ++local +isomalloc_sync`) with `AMPI Logic error: received late out-of-order message!`
> 
> I'll take a look

I assume that's the 64 bit build?

**Comment by evan-charmworks** [comment]
> I assume that's the 64 bit build?

Yes, we don't have a 32-bit CI target.

**Comment by matthiasdiener** [comment]
> > > netlrts aborted in megampi (`./charmrun ./pgm +p2 +vp2 ++local +isomalloc_sync`) with `AMPI Logic error: received late out-of-order message!`
> > 
> > 
> > I'll take a look
> 
> I assume that's the 64 bit build?

Yes, see here for the full log: https://circleci.com/gh/UIUC-PPL/charm/4151

**Comment by stwhite91** [comment]
Now the mpi-smp build is failing in megampi after migration: 

```
$ ./charmrun +p2 ./pgm +vp1  ++local

Running on 2 processors:  ./pgm +vp1 
charmrun> mpirun -np 2  ./pgm +vp1 
Charm++> Running on MPI version: 3.1
Charm++> level of thread support used: MPI_THREAD_FUNNELED (desired: MPI_THREAD_FUNNELED)
Charm++> Running in SMP mode: 2 processes, 1 worker threads (PEs) + 1 comm threads per process, 2 PEs total
Charm++> The comm. thread both sends and receives messages
Converse/Charm++ Commit ID: v6.10.0-rc2-145-g7c4546500
CharmLB> Load balancer assumes all CPUs are same.
Charm++> Running on 1 hosts (1 sockets x 4 cores x 2 PUs = 8-way SMP)
Charm++> cpu topology info is gathered in 0.000 seconds.
CharmLB> RandCentLB created.
[0] Testing: Testing...
[0] Testing: Testing...
[0] Testing: Testing...
[wirelessprv-10-195-127-93:93080] *** Process received signal ***
[wirelessprv-10-195-127-93:93080] Signal: Segmentation fault: 11 (11)
[wirelessprv-10-195-127-93:93080] Signal code: Address not mapped (1)
[wirelessprv-10-195-127-93:93080] Failing at address: 0x10c2ed860
[wirelessprv-10-195-127-93:93080] [ 0] 0   libsystem_platform.dylib            0x00007fff733be42d _sigtramp + 29
[wirelessprv-10-195-127-93:93080] [ 1] 0   ???                                 0x00007fe39c61f9d8 0x0 + 140615557970392
[wirelessprv-10-195-127-93:93080] [ 2] 0   pgm                                 0x000000010f0f94d3 _ZN6TCharm5startEv + 35
[wirelessprv-10-195-127-93:93080] [ 3] 0   pgm                                 0x000000010f0d73e5 _ZN6TCharm14ResumeFromSyncEv + 21
[wirelessprv-10-195-127-93:93080] [ 4] 0   pgm                                 0x000000010f21b094 _ZN12CkMigratable20staticResumeFromSyncEPv + 6036
[wirelessprv-10-195-127-93:93080] [ 5] 0   pgm                                 0x000000010f30a8b2 _ZN12LocalBarrier13ResumeClientsEv + 530
[wirelessprv-10-195-127-93:93080] [ 6] 0   pgm                                 0x000000010f30a68c _ZN10LBDatabase13ResumeClientsEv + 220
[wirelessprv-10-195-127-93:93080] [ 7] 0   pgm                                 0x000000010f36da76 _ZN9CentralLB13ResumeClientsEi + 38
[wirelessprv-10-195-127-93:93080] [ 8] 0   pgm                                 0x000000010f3767b7 _ZN17CkIndex_CentralLB30_call_ResumeClients_marshall11EPvS0_ + 183
[wirelessprv-10-195-127-93:93080] [ 9] 0   pgm                                 0x000000010f1e3de0 CkDeliverMessageFree + 64
[wirelessprv-10-195-127-93:93080] [10] 0   pgm                                 0x000000010f1e4dc1 _ZL19_invokeEntryNoTraceiP8envelopePv + 161
[wirelessprv-10-195-127-93:93080] [11] 0   pgm                                 0x000000010f1eae53 _ZL12_invokeEntryiP8envelopePv + 35
[wirelessprv-10-195-127-93:93080] [12] 0   pgm                                 0x000000010f1e8457 _ZL17_deliverForBocMsgP11CkCoreStateiP8envelopeP8IrrGroup + 135
[wirelessprv-10-195-127-93:93080] [13] 0   pgm                                 0x000000010f1e67d0 _ZL17_processForBocMsgP11CkCoreStateP8envelope + 176
[wirelessprv-10-195-127-93:93080] [14] 0   pgm                                 0x000000010f1e64bc _Z15_processHandlerPvP11CkCoreState + 428
[wirelessprv-10-195-127-93:93080] [15] 0   pgm                                 0x000000010f39acb5 CmiHandleMessage + 181
[wirelessprv-10-195-127-93:93080] [16] 0   pgm                                 0x000000010f39b134 CsdScheduleForever + 228
[wirelessprv-10-195-127-93:93080] [17] 0   pgm                                 0x000000010f39ad0a CsdScheduler + 26
[wirelessprv-10-195-127-93:93080] [18] 0   pgm                                 0x000000010f395b80 _ZL13ConverseRunPEi + 976
[wirelessprv-10-195-127-93:93080] [19] 0   pgm                                 0x000000010f397c77 _ZL12call_startfnPv + 71
[wirelessprv-10-195-127-93:93080] [20] 0   libsystem_pthread.dylib             0x00007fff733c9e65 _pthread_start + 148
[wirelessprv-10-195-127-93:93080] [21] 0   libsystem_pthread.dylib             0x00007fff733c583b thread_start + 15
```

**Comment by evan-charmworks** [comment]
I can't reproduce the MPI SMP issue, and when I re-ran that CI it finished successfully.

**Comment by stwhite91** [comment]
I'm able to reproduce the failure on my Macbook with an MPI-SMP build, and Eric B is still seeing a failure with this patch in adcirc on AMPI

**Comment by ericjbohm** [comment]
I can produce it on netlrts-linux-x86_64 in address sanitizer with +p1 +vp2.  Note, non-SMP, so this using only tcharm level threading.

**Comment by evan-charmworks** [comment]
> I'm able to reproduce the failure on my Macbook with an MPI-SMP build.

I can reproduce the failure you posted above but it is due to ASLR, not this PR. If you run with `++debug-no-pause` it won't occur. Here I ran with `++debug` and specified `settings set target.disable-aslr false`.

```
(lldb) target create "./pgm"
Current executable set to './pgm' (x86_64).
(lldb) settings set -- target.run-args  "+vp1"
(lldb) settings set target.disable-aslr false
(lldb) r
Process 33663 launched: '/Users/evan/charm/mpi-darwin-x86_64-smp/tests/ampi/megampi/pgm' (x86_64)
Process 33663 stopped
* thread #4, stop reason = EXC_BAD_ACCESS (code=1, address=0x1021e6e60)
    frame #0: 0x00000001021e6e60
error: memory read failed for 0x1021e6e00
Target 0: (pgm) stopped.
(lldb) bt
* thread #4, stop reason = EXC_BAD_ACCESS (code=1, address=0x1021e6e60)
  * frame #0: 0x00000001021e6e60
    frame #1: 0x000000010349baf4 pgm`::CthAwaken(th=0x00007f8ce84530e0) at threads.C:942:3
    frame #2: 0x00000001032398c3 pgm`TCharm::start(this=0x00007f8ce84523a0) at tcharm_impl.h:272:3
    frame #3: 0x000000010321ca35 pgm`TCharm::ResumeFromSync(this=0x00007f8ce84523a0) at tcharm.C:488:3
    frame #4: 0x0000000103349cdf pgm`CkMigratable::staticResumeFromSync(data=0x00007f8ce84523a0) at cklocation.C:1761:9
    frame #5: 0x0000000103412fdd pgm`LocalBarrier::ResumeClients(this=0x00007f8ce862a058) at LBDatabase.C:1398:5
    frame #6: 0x0000000103412f5c pgm`LBDatabase::ResumeClients(this=0x00007f8ce8629d40) at LBDatabase.C:1063:16
    frame #7: 0x00000001034777a6 pgm`CentralLB::ResumeClients(this=0x00007f8ce862a320, balancing=1) at CentralLB.C:1261:12
    frame #8: 0x000000010348042f pgm`CkIndex_CentralLB::_call_ResumeClients_marshall11(impl_msg=0x00007f8ce8452080, impl_obj_void=0x00007f8ce862a320) at CentralLB.def.h:1946:13
    frame #9: 0x0000000103309e30 pgm`::CkDeliverMessageFree(epIdx=199, msg=0x00007f8ce8452080, obj=0x00007f8ce862a320) at ck.C:551:3
    frame #10: 0x000000010330adc1 pgm`_invokeEntryNoTrace(epIdx=199, env=0x00007f8ce8452040, obj=0x00007f8ce862a320) at ck.C:606:5
    frame #11: 0x0000000103311f74 pgm`_invokeEntry(epIdx=199, env=0x00007f8ce8452040, obj=0x00007f8ce862a320) at ck.C:617:5
    frame #12: 0x000000010330e2d7 pgm`_deliverForBocMsg(ck=0x00007f8ce8628350, epIdx=199, env=0x00007f8ce8452040, obj=0x00007f8ce862a320) at ck.C:1069:3
    frame #13: 0x000000010330c9e0 pgm`_processForBocMsg(ck=0x00007f8ce8628350, env=0x00007f8ce8452040) at ck.C:1091:5
    frame #14: 0x000000010330c6cc pgm`_processHandler(converseMsg=0x00007f8ce8452040, ck=0x00007f8ce8628350) at ck.C:1262:7
    frame #15: 0x00000001034a61e5 pgm`::CmiHandleMessage(msg=0x00007f8ce8452040) at convcore.C:1693:2
    frame #16: 0x00000001034a66b4 pgm`::CsdScheduleForever() at convcore.C:1946:7
    frame #17: 0x00000001034a623a pgm`::CsdScheduler(maxmsgs=-1) at convcore.C:1879:17
    frame #18: 0x00000001034a115a pgm`ConverseRunPE(everReturn=0) at machine-common-core.C:1585:30
    frame #19: 0x00000001034a3527 pgm`call_startfn(vindex=0x0000000000000000) at machine-smp.C:379:3
    frame #20: 0x00007fff66d58e65 libsystem_pthread.dylib`_pthread_start + 148
    frame #21: 0x00007fff66d5483b libsystem_pthread.dylib`thread_start + 15
(lldb) f 1
frame #1: 0x000000010349baf4 pgm`::CthAwaken(th=0x00007f8ce84530e0) at threads.C:942:3
   939    B(th)->scheduled++;
   940    CthThreadToken * token = B(th)->token;
   941    constexpr int strategy = CQS_QUEUEING_FIFO;
-> 942    awakenfn(token, strategy, 0, 0); // If this crashes, disable ASLR.
   943    /*B(th)->scheduled = 1; */
   944    /*changed due to out-of-core emulation in BigSim */
   945 
```

**Comment by evan-charmworks** [comment]
Even if there is another issue that needs to be fixed, this PR might help in some cases.

**Comment by evan-charmworks** [comment]
I think we should merge this.

**Pull request merged**